Pr 2135

trxcllnt opened the pull request:
@jayphelps here are the changes I proposed earlier for `timeout` and `timeoutWith`. Had to fix a bug with rescheduled VirtualActions (these seem to be the first operators to take advantage of the fact that actions can be recycled), but other than that it should be pretty straightforward.

trxcllnt committed:
build(npm-scripts): update debug_mocha npm script for node 6

trxcllnt committed:
fix(VirtualAction): Block rescheduled VirtualActions from executing their scheduled work.

VirtualActions are immutable so they can be inspected by the TestScheduler. In order to mirror

rescheduled stateful Actions, rescheduled VirtualActions shouldn't execute if they've been

rescheduled before execution.

jayphelps commented on src/operator/timeout.ts:
```diff
@@ -65,56 +56,36 @@ class TimeoutSubscriber<T> extends Subscriber<T> {
+      // VirtualActions are immutable, so they create and return a clone. In this
+      // case, we need to set the action reference to the most recent VirtualAction,
+      // to ensure that's the one we clone from next time.
+      this.action = (<any> action).schedule(this, this.waitFor);
```
How bout we set the type `private action: Action` instead of `Subscription` so we don't need to cast? I didn't use Action to begin with because..well...at the time I didn't need it and just lazy.

trxcllnt replied:
We can, but we'll have to cast it to `Action<TimeoutSubscriber<T>>` each time. Both `Scheduler.schedule` and `Action.schedule` are typed to return `Subscription`.

jayphelps commented on src/operator/timeout.ts:
```diff
@@ -65,56 +56,36 @@ class TimeoutSubscriber<T> extends Subscriber<T> {
+  private _unsubscribe() {
+    this.action = null;
+    this.scheduler = null;
+    this.errorToSend = null;
```
are these null-outs necessary?

trxcllnt replied:
They _should_ be GC'd when the subscriber is dereferenced, but it's good practice for subscribers to clean up any references they have.

jayphelps commented on src/scheduler/VirtualTimeScheduler.ts:
```diff
@@ -77,6 +82,12 @@ export class VirtualAction<T> extends AsyncAction<T> {
+      return super._execute(state, delay);
+    }
+  }
+
```
I hate to ask...but would it be tough to get an isolated test for these scheduler changes? e.g. if some day no operators utilize this behavior, it's still tested by itself

jayphelps commented:
thanks dudebro! sexy stuff. Just a couple comments inline.

jayphelps commented:
@trxcllnt bump

trxcllnt commented:
@jayphelps sorry been busy I agree with your comments will aim to update this tonight.

trxcllnt committed:
fix(timeout): Update timeout and timeoutWith to recycle their scheduled timeout actions.

The timeout and timeoutWith operators should dispose their scheduled timeout actions on

unsubscription. Also, given the new scheduling architecture, they can recycle their scheduled

actions so just one action is allocated per subscription.

#2134 #2135
